source/common/network/lc_trie.h

```diff
@@ -0,0 +1,488 @@
+     *
+     * Note: This trie handles Ipv4 and Ipv6 separately.
+     * Note: The trie can only support up 524288(2^19) nodes. Please see LcNode for further
+     * explanation.
+     *
+     * @param fill_factor supplies the fraction of completeness to use when considering
+     *        if a subtrie should be compressed.
+     * @param root_branching_factor supplies the branching factor at the root. This is
+     *        an optimization as suggested in the paper.
+     * @param tag_data supplies the ip tag and the prefixes for associated with that tag.
+     */
+    LcTrieInternal(std::vector<IpPrefixes<IpType>>& tag_data, double fill_factor = 0.5,
+                   uint32_t root_branching_factor = 0);
+
+    /**
+     * Search the LC Trie for a prefix that has the input, ip, in range.
+     * @param  ip supplies the IP address searching for a prefix in string format.
+     * @return tag from the prefix that encompasses the input. An empty string is returned
+     *         if no prefix in the LC Trie exists.
+     */
+    std::string search(const IpType& ip_address) const;
+
+  private:
+    /**
+     * Builds the Level Compresesed Trie, by first sorting the base vector, removing nested prefixes
+     * and duplicated and then recursively builds the trie by invoking buildRecursively().
+     */
+    void buildLcTrie(std::vector<IpPrefixes<IpType>>& tag_data) {
+      if (tag_data.empty()) {
+        return;
+      }
+
+      // The node used for the trie uses the last 20 bits of uint32_t. The trie cannot support
+      // an input greater than 524288(2^19) nodes.
+      // TODO(ccaraman): Add a test case for this.
+      if (tag_data.size() > (1 << 19)) {
+        throw EnvoyException(fmt::format(
+            "Size of the input vector({0}) for LC-Trie is larger than the max size supported(2^19)",
+            tag_data.size()));
+      }
+
+      std::sort(tag_data.begin(), tag_data.end());
+      ip_prefixes_.push_back(tag_data[0]);
+
+      // Remove duplicate entries and nested prefixes.
+      for (size_t i = 1; i < tag_data.size(); ++i) {
+        // TODO(ccaraman): Add support for nested prefixes.
+        if (tag_data[i - 1].isPrefix(tag_data[i])) {
+          // TODO(ccaraman): Add printing of nested prefixes to simplify debugging.
+          throw EnvoyException(
+              fmt::format("LcTrie does not support nested prefixes."));
+        }
+        if (tag_data[i - 1].compare(tag_data[i]) != 0) {
+          ip_prefixes_.push_back(tag_data[i]);
+        }
+      }
+
+      // The trie_ vector can have at most twice the number of nodes in the ip_prefixes vector.
+      // However, due to the fill factor a buffer is added. Preallocate the size of the vector to
+      // avoid resizing while building the trie.
+      // TODO(ccaraman): Define a better buffer value for the trie_ size.
+      // TODO(ccaraman): Define a better buffer value for the trie_ size.
+      // The value(2000000) was used in the original implementation in
```

Comment:
@brian-pane Do you have any insight as to why `2000000` was chosen? Is there a better way to compute a buffer given the fill_factor?

Reply:
In the original implementation, I think the author just picked 2000000 as a really big number that they thought would be big enough to handle any situation.  And then, after populating the buffer, they resized it down to the amount of space they actually ended up using.

I don't have a good way to compute the buffer size in advance, since it depends on the exact shape of the tree. However, I think it's okay to _not_ pre-size the buffer. I think std::vector uses a power-of-two allocator internally, so the number of resize operations will be O(log(N)) for N prefixes.  So my proposal is to just use std::vector, see how fast it is, and then optimize if needed.

Reply:
That sounds good! I prefer your suggestion to having a magic number. I'll change the comment to say how this diverged from the reference implementation.

Reply:
I decided to not remove the resizing due to the accessing pattern used across the code base. Once there are performance benchmark tests, the resizing the vector can be something we look to improve if needed.